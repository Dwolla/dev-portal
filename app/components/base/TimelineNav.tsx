import styled from "@emotion/styled";
import { ReactComponent as CheckMarkIcon } from "../../../assets/images/component-icons/check-mark.svg";
import { GREY_3, CODE_BLOCK_ORANGE, PURPLE_DARK } from "../colors";
import { POPPINS } from "../typography";

const StyledDiv = styled.div`
  display: flex;
  justify-content: space-between;
  width: 75%;
  cursor: pointer;
`;

const StyledStepNum = styled.div`
  width: 25px;
  height: 25px;
  font-family: ${POPPINS};
  font-style: normal;
  font-weight: normal;
  font-size: 14px;
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid ${GREY_3};
  color: ${GREY_3};
  border-radius: 50%;
  margin: 14px 7px 9px;
  &.active {
    border-color: ${CODE_BLOCK_ORANGE};
    color: ${CODE_BLOCK_ORANGE};
  }
  &.completed {
    border-color: ${CODE_BLOCK_ORANGE};
    color: ${CODE_BLOCK_ORANGE};
  }
`;

const StyledText = styled.div`
  font-family: ${POPPINS};
  text-align: center;
  font-style: normal;
  font-weight: normal;
  font-size: 12px;
  width: 105px;
  color: ${GREY_3};
  &.active {
    color: ${PURPLE_DARK};
  }
  &.completed {
    color: ${PURPLE_DARK};
  }
`;

const StyledStep = styled.div`
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  flex: 1;
  &:after {
    position: absolute;
    content: "";
    border: 1px solid ${GREY_3};
    width: 77%;
    top: 25px;
    left: 62%;
  }
  &:last-child {
    &:after {
      content: none;
    }
  }
`;

const createElements = (steps, active, setActive) => {
  const onClick = (e) => {
    if (e.target.id) setActive(parseInt(e.target.id, 10));
  };
  const elements = steps.map((step, idx) => {
    let status = null;
    let num = null;
    if (idx === active) {
      status = "active";
      num = idx;
    } else if (idx < active) {
      status = "completed";
      num = (
        <CheckMarkIcon
          id={idx}
          width={14}
          height={10}
          stroke={CODE_BLOCK_ORANGE}
        />
      );
    } else {
      num = idx;
    }
    return (
      <StyledStep className={status} key={step}>
        <StyledStepNum
          className={status}
          onClick={idx !== 0 ? onClick : undefined}
          id={idx}
        >
          {num}
        </StyledStepNum>
        <StyledText className={status}>{step}</StyledText>
      </StyledStep>
    );
  });
  return elements;
};

type Props = {
  totalSteps: Array<string>;
  active: number;
  setActive: Function;
};

function TimelineNav({ totalSteps, active, setActive }: Props) {
  return (
    <div style={{ display: "flex", justifyContent: "center" }}>
      <StyledDiv>{createElements(totalSteps, active, setActive)}</StyledDiv>
    </div>
  );
}

export default TimelineNav;
