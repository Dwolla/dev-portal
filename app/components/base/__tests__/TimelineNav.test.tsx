import renderer from "react-test-renderer";
import TimelineNav from "../TimelineNav";

const steps = ["Create Transfer", "First Leg", "Second Leg", "Third Leg"];

test("TimelineNav", () => {
  const setState = jest.fn();
  const tree = renderer
    .create(<TimelineNav totalSteps={steps} active={0} setActive={setState} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
