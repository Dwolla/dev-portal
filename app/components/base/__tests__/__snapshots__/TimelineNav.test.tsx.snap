// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineNav 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "justifyContent": "center",
    }
  }
>
  <div
    className="css-ec7nrl-StyledDiv e92fyip3"
  >
    <div
      className="active css-zr8nux-StyledStep e92fyip0"
    >
      <div
        className="active css-e0ruun-StyledStepNum e92fyip2"
        id={0}
      >
        0
      </div>
      <div
        className="active css-1g7t0p2-StyledText e92fyip1"
      >
        Create Transfer
      </div>
    </div>
    <div
      className="css-zr8nux-StyledStep e92fyip0"
    >
      <div
        className="css-e0ruun-StyledStepNum e92fyip2"
        id={1}
        onClick={[Function]}
      >
        1
      </div>
      <div
        className="css-1g7t0p2-StyledText e92fyip1"
      >
        First Leg
      </div>
    </div>
    <div
      className="css-zr8nux-StyledStep e92fyip0"
    >
      <div
        className="css-e0ruun-StyledStepNum e92fyip2"
        id={2}
        onClick={[Function]}
      >
        2
      </div>
      <div
        className="css-1g7t0p2-StyledText e92fyip1"
      >
        Second Leg
      </div>
    </div>
    <div
      className="css-zr8nux-StyledStep e92fyip0"
    >
      <div
        className="css-e0ruun-StyledStepNum e92fyip2"
        id={3}
        onClick={[Function]}
      >
        3
      </div>
      <div
        className="css-1g7t0p2-StyledText e92fyip1"
      >
        Third Leg
      </div>
    </div>
  </div>
</div>
`;
